contracts/drafts/ERC777/ERC777.sol

```diff
@@ -0,0 +1,402 @@
+     * @param operatorData bytes extra information provided by the operator (if any)
+     */
+    function _send(
+        address operator,
+        address from,
+        address to,
+        uint256 amount,
+        bytes memory userData,
+        bytes memory operatorData
+    )
+    private
+    {
+        require(from != address(0));
+        require(to != address(0));
+        require((amount % _granularity) == 0);
+
+        _callTokensToSend(operator, from, to, amount, userData, operatorData);
+
+        // Update state variables
+        _balances[from] = _balances[from].sub(amount);
+        _balances[to] = _balances[to].add(amount);
+        assert((_balances[from] % _granularity) == 0);
+        assert((_balances[to] % _granularity) == 0);
+
+        _callTokensReceived(operator, from, to, amount, userData, operatorData);
+
+        emit Sent(operator, from, to, amount, userData, operatorData);
+    }
+
+    /**
+     * @dev Call from.tokensToSend() if the interface is registered
+     * @param operator address operator requesting the transfer
+     * @param from address token holder address
+     * @param to address recipient address
+     * @param amount uint256 amount of tokens to transfer
+     * @param userData bytes extra information provided by the token holder (if any)
+     * @param operatorData bytes extra information provided by the operator (if any)
+     */
+    function _callTokensToSend(
+        address operator,
+        address from,
+        address to,
+        uint256 amount,
+        bytes memory userData,
+        bytes memory operatorData
+    )
+    private
+    {
+        address implementer = _erc1820.getInterfaceImplementer(from, TOKENS_SENDER_INTEFACE_HASH);
+        if (implementer != address(0)) {
+            IERC777Sender(implementer).tokensToSend(operator, from, to, amount, userData, operatorData);
+        }
+    }
+
+    /**
+    * @dev Call to.tokensReceived() if the interface is registered
+    * @param operator address operator requesting the transfer
+    * @param from address token holder address
+    * @param to address recipient address
+    * @param amount uint256 amount of tokens to transfer
+    * @param userData bytes extra information provided by the token holder (if any)
+        * @param operatorData bytes extra information provided by the operator (if any)
+            * @return false if the recipient is a contract but tokensReceived() was not
+                *         registered for the recipient
+                    */
+    function _callTokensReceived(
+        address operator,
+        address from,
+        address to,
+        uint256 amount,
+        bytes memory userData,
+        bytes memory operatorData
+    )
+    private
+    returns(bool)
+    {
+        address implementer = _erc1820.getInterfaceImplementer(to, TOKENS_RECIPIENT_INTERFACE_HASH);
+        if (implementer == address(0)) {
+            return(!to.isContract());
```

Comment:
From my understanding of the following bit of the EIP, we should `require` the value that this function returns (which is done on `_mint` but not `_send`):

>The token contract MUST revert when sending in any of the following cases:
    * The recipient is a contract, and it does not implement the ERC777TokensRecipient interface via ERC1820.